src/libopensc/card-dtrust.c

```diff
@@ -0,0 +1,456 @@
+		card->type = SC_CARD_TYPE_DTRUST_V4_4_STD;
+	else if(plen >= 28 && !memcmp(pp, "D-TRUST Card 4.4 Multi ECC 2", 28))
+		card->type = SC_CARD_TYPE_DTRUST_V4_4_MULTI;
+	else
+		return SC_ERROR_WRONG_CARD;
+
+	name = malloc(plen + 1);
+	if(name == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+	memcpy(name, pp, plen);
+	name[plen] = '\0';
+	card->name = name;
+
+	sc_log(card->ctx, "found %s", card->name);
+
+	return SC_SUCCESS;
+}
+
+static int dtrust_match_card(sc_card_t *card)
+{
+	if(_sc_match_atr(card, dtrust_atrs, &card->type) < 0)
+		return 0;
+
+	if(_dtrust_match_cardos(card) != SC_SUCCESS)
+		return 0;
+
+	if(_dtrust_match_profile(card) != SC_SUCCESS)
+		return 0;
+
+	sc_log(card->ctx, "D-Trust Signature Card (CardOS 5.4)");
+
+	return 1;
+}
+
+static int _dtrust_get_serialnr(sc_card_t *card)
+{
+	int r;
+	u8 buf[32];
+
+	r = sc_get_data(card, 0x0181, buf, 32);
+	LOG_TEST_RET(card->ctx, r, "querying serial number failed");
+
+	if(r != 8)
+	{
+		sc_log(card->ctx, "unexpected response to GET DATA serial number");
+		return SC_ERROR_INTERNAL;
+	}
+
+	/* cache serial number */
+	memcpy(card->serialnr.value, buf, 8);
+	card->serialnr.len = 8;
+
+	return SC_SUCCESS;
+}
+
+static int dtrust_init(sc_card_t *card)
+{
+	int r;
+	const size_t data_field_length = 437;
+	unsigned long flags, ext_flags;
+
+	SC_FUNC_CALLED(card->ctx, SC_LOG_DEBUG_VERBOSE);
+
+	card->cla = 0x00;
+
+	card->drv_data = calloc(1, sizeof(struct dtrust_drv_data_t));
+	if(card->drv_data == NULL)
+		return SC_ERROR_OUT_OF_MEMORY;
+
+	r = _dtrust_get_serialnr(card);
+	LOG_TEST_RET(card->ctx, r, "Error reading serial number.");
+
+	card->caps |= SC_CARD_CAP_APDU_EXT | SC_CARD_CAP_ISO7816_PIN_INFO;
+
+	card->max_send_size = data_field_length - 6;
+#ifdef _WIN32
+	/* see card-cardos.c */
+	if(card->reader->max_send_size == 255 && card->reader->max_recv_size == 256)
+	{
+		sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE, "resetting reader to use data_field_length");
+		card->reader->max_send_size = data_field_length - 6;
+		card->reader->max_recv_size = data_field_length - 3;
+	}
+#endif
+
+	card->max_send_size = sc_get_max_send_size(card); /* see card-cardos.c */
+	card->max_recv_size = data_field_length - 2;
+	card->max_recv_size = sc_get_max_recv_size(card);
+
+	flags = 0;
+	r = SC_ERROR_WRONG_CARD;
+
+	switch(card->type)
+	{
+	case SC_CARD_TYPE_DTRUST_V4_1_STD:
+	case SC_CARD_TYPE_DTRUST_V4_4_STD:
+		flags |= SC_ALGORITHM_RSA_PAD_PKCS1;
+		flags |= SC_ALGORITHM_RSA_PAD_PSS;
+		flags |= SC_ALGORITHM_RSA_PAD_OAEP;
+		flags |= SC_ALGORITHM_RSA_HASH_SHA256;
+		flags |= SC_ALGORITHM_RSA_HASH_SHA384;
+		flags |= SC_ALGORITHM_RSA_HASH_SHA512;
+		flags |= SC_ALGORITHM_MGF1_SHA256;
+		flags |= SC_ALGORITHM_MGF1_SHA384;
+		flags |= SC_ALGORITHM_MGF1_SHA512;
+
+		_sc_card_add_rsa_alg(card, 3072, flags, 0);
+
+		r = SC_SUCCESS;
+		break;
+
+	/* Untested due to lacking hardware */
+	case SC_CARD_TYPE_DTRUST_V4_1_MULTI:
+	case SC_CARD_TYPE_DTRUST_V4_1_M100:
+	case SC_CARD_TYPE_DTRUST_V4_4_MULTI:
```

Comment:
Do I understand it right that the first two tested cards support RSA only and these three (untested) support EC only?

Reply:
Yes, your are right. It's specified in this way by the vendor, but I have no idea why the so called multi-cards use elliptic curves instead of RSA. From the users point of view the only difference between the cards is, that the multi-cards don't require a PIN verification for each single signature.

I am currently checking if my employer can obtain a multi-card for me, so that I can implement this type of card as well.